src/util/hashFetch/index.js

```diff
@@ -0,0 +1,247 @@
+// Copyright 2015-2017 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+const fs = require('fs');
+const util = require('util');
+const path = require('path');
+
+import unzip from 'unzipper';
+
+import { getHashFetchPath } from '../host';
+import ExpoRetry from './expoRetry';
+import Contracts from '@parity/shared/lib/contracts';
+import { sha3 } from '@parity/api/lib/util/sha3';
+import { bytesToHex } from '@parity/api/lib/util/format';
+import { ensureDir as fsEnsureDir, emptyDir as fsEmptyDir } from 'fs-extra';
+import { http, https } from 'follow-redirects';
+
+const fsExists = util.promisify(fs.stat);
+const fsStat = util.promisify(fs.stat);
+const fsRename = util.promisify(fs.rename);
+const fsReadFile = util.promisify(fs.readFile);
+const fsUnlink = util.promisify(fs.unlink);
+const fsReaddir = util.promisify(fs.readdir);
+const fsRmdir = util.promisify(fs.rmdir);
+
+const MAX_DOWNLOADED_FILE_SIZE_BYTES = 10485760; // 20MB
+
+function registerFailedAttemptAndThrow (hash, url, e) {
+  ExpoRetry.get().registerFailedAttempt(hash, url);
+  throw e;
+}
+
+function checkHashMatch (hash, path) {
+  return fsReadFile(path).then(content => {
```

Comment:
This will read the whole file to memory, might be better to just hash without allocating. In general we should avoid reading the files that were fetched over the network, better to just stream them if it's possible.